data_structures/linked_list/deque_doubly.py

```diff
@@ -0,0 +1,144 @@
+        self._trailer._prev = self._header
+        self._size = 0
+    
+    def __len__(self):
+        return self._size
+    
+    def is_empty(self):
+        return self.__len__() == 0
+    
+    def _insert(self, predecessor, e, successor):
+        # Create new_node by setting it's prev.link -> header
+        # setting it's next.link -> trailer
+        new_node = self._Node(predecessor, e, successor)
+        predecessor._next = new_node
+        successor._prev = new_node
+        self._size += 1
+        return self
+    
+    def _delete(self, node):
+        predecessor = node._prev
+        successor = node._next
+        
+        predecessor._next = successor
+        successor._prev = predecessor
+        self._size -= 1
+        temp = node._data
+        node._prev = node._next = node._data = None
+        del node
+        return temp
+
+class LinkedDeque(_DoublyLinkedBase):
+    
+    def first(self):
+        """ return first element 
+        >>> d = LinkedDeque()
+        >>> d.add_first('A') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.first()
+        'A'
+        >>> d.add_first('B') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.first()
+        'B'
+        """ 
+        if self.is_empty():
+            raise Exception('List is empty')
+        return self._header._next._data
+    
+    def last(self):
+        """ return last element
+        >>> d = LinkedDeque()
+        >>> d.last()
+        Traceback (most recent call last):
+           ...
+        Exception: List is empty
+        >>> d.add_last('A') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.last()
+        'A'
+        >>> d.add_last('B') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        >>> d.last()
+        'B'
+        """
+        if self.is_empty():
+            raise Exception('List is empty')
+        return self._trailer._prev._data
+    
+    ### DEque Insert Operations (At the front, At the end) ###
+    
+    def add_first(self, element):
+        """ insertion in the front
+        >>> d = LinkedDeque()
+        >>> d.add_first('AV') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        """
+        return self._insert(self._header, element, self._header._next)
+    
+    def add_last(self, element):
+        """ insertion in the end
+        >>> d = LinkedDeque()
+        >>> d.add_last('A') # doctest: +ELLIPSIS
+        <deque_doubly.LinkedDeque object at ...
+        """
+        return self._insert(self._trailer._prev, element, self._trailer)
+    
+    ### DEqueu Remove Operations (At the front, At the end) ###
+    
+    def remove_first(self):
+        """ removal from the front
+        >>> d = LinkedDeque()
+        >>> d.remove_first()
+        Traceback (most recent call last):
+           ...
+        Exception: List is empty
```

Comment:
This should be the same exception and a similar error message as __[].pop()__ --> IndexError: pop from empty list

So something like __IndexError: remove_first from empty list__